Zend/zend_execute_API.c

```diff
@@ -830,6 +830,12 @@ int zend_call_function(zend_fcall_info *fci, zend_fcall_info_cache *fci_cache) /
 			/* We must re-initialize function again */
 			fci_cache->function_handler = NULL;
 		}
+
+		/* This flag is regularly checked while running user functions, but not internal
+		 * So see whether we timed out while the function was running... */
+		if (EG(timed_out)) {
+			zend_timeout();
+		}
```

Comment:
After thinking about this a bit, I think instead of checking timed_out specifically here, we should do a VM interrupt check as in https://github.com/php/php-src/blob/5d93eab6207614d44378f0034656cf382963244b/Zend/zend_vm_def.h#L9206. After all the core issue here applies not just to internal timeouts, but also pcntl signal handling, I think.

Reply:
OK. I think I have figured out what you mean...

`pcntl` sets a signal handler which sets `vm_interrupt` to 1. That causes execution to bounce into `pcntl`'s `zend_interrupt_function` next time the VM checks the interrupt flag. But that only happens when the VM is running PHP bytecode.

If you are running native shutdown functions, and `pcntl` is enabled, and a signal comes along, it will set the `vm_interrupt` flag but `pcntl`'s `zend_interrupt_function` will never be called.

Do you think `pcntl` really should care about what is happening during shutdown, though? Maybe?

Reply:
> Do you think pcntl really should care about what is happening during shutdown, though? Maybe?

During shutdown probably isn't important, but there are other cases where we could run into this, at least in theory. For example, consider something like `array_map('expensive_internal_function', $array)`. This would allow a signal handler to be executed in the middle of the array_map, rather than at the end.

Maybe it's possible to test this with a combination of `InfiniteIterator` and `iterator_apply`.

Reply:
Aha. Wow, I never thought of that. OK.

Reply:
I've added an extra test in https://github.com/php/php-src/pull/5543/commits/7f35610dc9636485c95b1e630784f94d6874db0c. I was hoping to show that this leaks the time_nanosleep return value (which should be an array in this case), but due to the way array_map is implemented this doesn't happen...

I think we should be repeating the `EG(exception)` check from a few lines above if the VM interrupt is executed, to make sure that code not freeing the retval on exception doesn't leak...